src/style-spec/expression/index.js

```diff
@@ -60,70 +124,75 @@ function createExpression(expression: mixed,
         return error(parser.errors);
     }
 
-    const evaluator = new EvaluationContext();
-
-    let evaluate;
     if (options.handleErrors === false) {
-        evaluate = function (globals, feature) {
-            evaluator.globals = globals;
-            evaluator.feature = feature;
-            return parsed.evaluate(evaluator);
-        };
+        return success(new StyleExpression(parsed, propertySpec));
     } else {
-        const warningHistory: {[key: string]: boolean} = {};
-        const defaultValue = getDefaultValue(propertySpec);
-        let enumValues;
-        if (propertySpec.type === 'enum') {
-            enumValues = propertySpec.values;
+        return success(new StyleExpressionWithErrorHandling(parsed, propertySpec));
+    }
+}
+
+class ZoomConstantExpression<Kind> {
```

Comment:
I'd prefer to convert `ConstantExpression`, `SourceExpression`, etc. to classes, remove `kind`, and use `instanceof` for refinement.

Reply:
I considered this, but backed off from it mostly because normalizing stop functions to expressions would be sort of weird -- i.e., would we have `class XxxFunction extends XxxExpression`?   (Also, though more minor, `CameraExpression` and `CompositeExpression` would be identical (but distinct) classes)